common/src/main/java/io/pravega/common/util/BlockingAsyncIterator.java

```diff
@@ -0,0 +1,68 @@
+/**
+ * Copyright (c) 2017 Dell Inc., or its subsidiaries. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.pravega.common.util;
+
+import com.google.common.annotations.VisibleForTesting;
+import lombok.Synchronized;
+
+import javax.annotation.concurrent.ThreadSafe;
+import java.util.Iterator;
+import java.util.NoSuchElementException;
+
+/**
+ * This class is a wrapper over async iterator that implements java.util.Iterator interface.
+ * This is blocking wrapper on async iterator's getNext method which implements blocking hasNext and next methods of 
+ * iterator interface. 
+ * This class is threadsafe but that has the drawback of blocking on asyncIterator's getNext within a lock. 
+ */
+@ThreadSafe
+@VisibleForTesting
+public class BlockingAsyncIterator<T> implements Iterator<T> {
+    private final AsyncIterator<T> asyncIterator;
+    private T next;
+    private boolean canHaveNext;
+
+    public BlockingAsyncIterator(AsyncIterator<T> asyncIterator) {
+        this.canHaveNext = true;
+        this.next = null;
+        this.asyncIterator = asyncIterator;
+    }
+
+    @Synchronized
+    private void load() {
+        if (next == null && canHaveNext) {
+            next = asyncIterator.getNext().join();
```

Comment:
Waiting on a future while holding a lock is never a good idea. Here's my suggestion for avoiding this:
- make `next` a `CompletableFuture<T>` and still guard it by a lock
- In this method
    - While holding the lock, check `next` for null, and if so, then set it to `asyncInterator.getNext()` (no `join`)
    - After releasing the lock, invoke `join` on `next` and inspect its result.

You may have to do a bit more bookkeeping (in this method) to track whether you actually invoked `getNext` or not.

I think you can also get rid of `canHaveNext`. If `next` is completed (it should always be once this method exits), if its contents is `null`, then `hasNext()` should return `false` and `next()` should throw.

Reply:
Actually this class is supposed to implement java.util Iterator interface that we can give out to applications so they know what to do with it.
It’s used in streamManager class.

That’s why it has canhavenext implementation and all the future.join under a lock